File: scripts/generateLeetCodeContestHistory.mjs
import * as A from '../src/views/91/contest.js'
import axios from 'axios'
import fs from 'fs'
const TIMEOUT = 30 * 1000

const { getAllContext } = A

function sleep(ms) {
    return new Promise(r => {
        setTimeout(r, ms)
    })
}

async function getByPageNo(id, pageNo) {
    const url = `https://leetcode.com/contest/api/ranking/weekly-contest-${id}/?pagination=${pageNo}`
    // await sleep(500)
    return axios.get(url).then(res => res.data).then(async (res) => {
        if (!res.total_rank || res.total_rank.length == 0) {
            return []
        }
        console.log(`process pageNo: ${pageNo} with count ${res.total_rank.length}`)
        // return res.total_rank.concat(await getByPageNo(id, pageNo + 1))
        weekly.push(res.total_rank)
        getByPageNo(id, pageNo + 1)
    })
}
let weekly = []
async function start() {
    const { contests } = getAllContext(new Date('2022-04-03T10:30:00.000+08:00').getTime(), 287, new Date('2022-06-30T10:30:00.000+08:00').getTime())
    const ans = JSON.parse(fs.readFileSync("src/views/91/db/contest.json", { encoding: 'utf8' }))
    for (const contest of contests) {
        if (contest.passed == false) {
            fs.writeFileSync("src/views/91/db/upcommingContest.json", JSON.stringify(contest))
            break
        }
        const fp = `src/views/91/db/weekly-${contest.id}.json`
        try {
            weekly = JSON.parse(fs.readFileSync(fp, { encoding: 'utf8' }))
        } catch (e) { }
        // 断点续下载
        const ranks = await Promise.race([getByPageNo(contest.id, Math.floor(weekly.length / 25) + 1), sleep(TIMEOUT)])
        // weekly = weekly.concat(ranks)
        fs.writeFileSync(fp, JSON.stringify(weekly))
        for (const rank of weekly) {
            // 664 是 weekly-287 
            const username = rank.username
            if (!(username in ans)) {
                ans[rank.username] = [
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null,
                    null
                ]
            }
            ans[rank.username][rank.contest_id - 377 - contests[0].id] = rank
        }
    }
    return ans

}
const a = await start();
fs.writeFileSync("src/views/91/db/contest.json", JSON.stringify(a))

File: scripts/generateLeetCodeContestHistory.py
#!/usr/bin/env python3

import argparse
import datetime
import json
import os
import requests
import sys


def getRanking(contest):
    API_URL_FMT = (
        "https://leetcode.com/contest/api/ranking/weekly-contest-{}/?pagination={}&region=global"
    )
    page = 30
    persistent_file = "src/views/91/db//weekly-{}.json".format(contest)
    total_rank = json.load(open(persistent_file, "r"))
    retry_cnt = 0
    while page < 31:
        try:
            url = API_URL_FMT.format(contest, page)
            resp = requests.get(url).json()
            page_rank = resp["total_rank"]
            if 0 == len(page_rank):
                break
            total_rank.extend(resp["total_rank"])
            print("Retrieved ranking from page {}. {} retrieved.".format(page, len(total_rank)))
            page += 1
            retry_cnt = 0
        except:
            print(f"Failed to retrieved data of page {page}...retry...{retry_cnt}")
            retry_cnt += 1

    # discard and transform fields
    for rank in total_rank:
        rank.pop("contest_id", None)
        rank.pop("user_slug", None)
        rank.pop("country_code", None)
        rank.pop("global_ranking", None)

    print("Save retrieved ranking to {}".format(persistent_file))
    with open(persistent_file, "w") as fp:
        json.dump(total_rank, fp)


def getContestInfo(contest):
    def unSlug(slug):
        return " ".join([w.capitalize() for w in slug.split("-")])

    def isNew(contests, newContest):
        for c in contests:
            if newContest["title"] == c["title"]:
                return False
        return True

    while True:
        try:
            CONTEST_INFO_API_URL_FMT = "https://leetcode.com/contest/api/info/weekly-contest-{}/"
            resp = requests.get(CONTEST_INFO_API_URL_FMT.format(contest)).json()
            startTimestamp = int(resp["contest"]["start_time"])
            break
        except:
            print("Failed to retrieved contest info...retry...")

    newContest = {"title": unSlug(contest), "slug": contest, "startTime": startTimestamp}
    contest_py = "src/views/91/db/contests-py.json"
    if os.path.exists(contest_py):
        with open(contest_py, "r") as fp:
            contests = json.load(fp)
    else:
        contests = []

    if isNew(contests, newContest):
        contests.append(newContest)
        contests.sort(key=lambda c: c["startTime"], reverse=True)

    with open(contest_py, "w+") as fp:
        json.dump(contests, fp)


def main():
    parser = argparse.ArgumentParser(description="Leetcode ranking crawler")
    parser.add_argument("contest", help="contest slug (ex: weekly-contest-178)")
    args = parser.parse_args()

    getRanking(args.contest)
    getContestInfo(args.contest)


if __name__ == "__main__":
    main()